### ISSUE TITLE
Export: when linking meshes with the same name, exporter can't tell them apart

### ISSUE THREAD

**Issue by scurest** [KhronosGroup/glTF-Blender-IO#1211]
Found in #1210.

If you have two meshes with the same name (from linking), the exporter thinks they are the same. Like I said in #1210, the library parameter to gather_mesh should presumably distinguish these since they have different `mesh.library`s, but the library parameter comes from `object.library`, not `mesh.library`.

**To Reproduce**
1. Create an empty .blend
2. Add > Cube
3. Rename the mesh "Cube" to "M"
4. Save as A.blend
5. Create an empty .blend
6. Add > UVSphere
7. Rename the mesh "Sphere" to "M"
8. File > Link and select A.blend/Mesh/M
9. Export glTF

![out](https://user-images.githubusercontent.com/11024420/93151203-e25b2800-f6c0-11ea-8e1d-89c3c9cb9996.png)

**Comment by scurest** [comment]
Using the library from `blender_object.data` fixes this, but I don't know if it breaks anything else.

<details>

<summary>Patch</summary>

````diff
diff --git a/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_nodes.py b/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_nodes.py
index 1a1abcd..2f786d2 100644
--- a/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_nodes.py
+++ b/addons/io_scene_gltf2/blender/exp/gltf2_blender_gather_nodes.py
@@ -326,7 +326,7 @@ def __gather_mesh(blender_object, library, export_settings):
                 blender_object_for_skined_data = blender_object
 
     result = gltf2_blender_gather_mesh.gather_mesh(blender_mesh,
-                                                   library,
+                                                   blender_object.data.library.name if blender_object.data.library else None,
                                                    blender_object_for_skined_data,
                                                    vertex_groups,
                                                    modifiers,
@@ -368,7 +368,7 @@ def __gather_mesh_from_nonmesh(blender_object, library, export_settings):
         blender_object_for_skined_data = None
 
         result = gltf2_blender_gather_mesh.gather_mesh(blender_mesh,
-                                                       library,
+                                                       blender_object.data.library.name if blender_object.data.library else None,
                                                        blender_object_for_skined_data,
                                                        vertex_groups,
                                                        modifiers,
````

</details>

**Comment by scurest** [comment]
The library for `blender_object_for_skinned_data` is also ignored.

**Comment by scurest** [comment]
I think it would be easier to handle the caching manually, rather than with the `@cached` stuff.

**Comment by scurest** [comment]
There's the same problem with gather_material where it caches based on the material name only. So I think the `__by_name` stuff in `cached` needs to go away: it we cache based on the regular `__hash__()` we can automatically tell apart different things with the same name.

However we don't want to cache on a temporary mesh like can be passed to gather_mesh, so gather_mesh can use a manually implemented rather than `@cached`.

**Comment by scurest** [comment]
gather_mesh also caches based on materials even if material are turned off.

**Comment by julienduroure** [comment]
Fixed by 69746b2b31eeee080d5dfd3c74f1e02742795ed2
